compiler/types.nim

```diff
@@ -1493,3 +1493,18 @@ proc typeMismatch*(conf: ConfigRef; info: TLineInfo, formal, actual: PType) =
       of efLockLevelsDiffer:
         msg.add "\nlock levels differ"
     localError(conf, info, msg)
+
+proc isTupleRecursive*(t: PType, cycleDetector: var IntSet): bool =
+  if t == nil: return
+  case t.kind:
+    of tyTuple:
+      if cycleDetector.containsOrIncl(t.id):
+        return true
+      for i in  0..<t.len:
+        var copy = cycleDetector
+        if isTupleRecursive(t[i], copy):
+          return true
+    of tyAlias, tyRef, tyPtr, tyGenericInst, tyVar, tyLent, tySink:
```

Comment:
Whitelisting seems dangerous, what if I try to create the recursion via ``tyArray`` or ``tySequence``?

Reply:
It is dangerous, see #10565 for an example of undetected recursion via `tySequence`.

Reply:
I looked into it, and yes ``tyArray``, ``tyUncheckedArray`` and ``tySequence`` is missing. But it is not dangerous to use white listing here at all. If the logic is wrong (which is already the case with the current implementation) illegal recursive types will be allowed at this stage. But signature hashing will eventually crash the compiler. In my opinion that is just some form of a badly styled error message.